File: conformance/bundles/22-reference-ambiguous-anchor/LibrarySteps.java
package dev.varar.conformance.bundle22;

import dev.varar.State;
import dev.varar.StepDefinitions;
import dev.varar.Steps;

/** Java sibling of {@code library.steps.ts} / {@code library.steps.py} (bundle {@code 22-reference-ambiguous-anchor}). */
public final class LibrarySteps implements StepDefinitions<LibrarySteps.Ctx> {

    record Ctx(int shelf) implements State {}

    @Override
    public void register(Steps<Ctx> s) {
        s.state(() -> new Ctx(0));

        s.stimulus("I shelve {int} books", (Ctx ctx, Integer n) -> new Ctx(ctx.shelf() + n));

        s.stimulus("I borrow a book", (Ctx ctx) -> new Ctx(ctx.shelf() - 1));

        s.sensor("The shelf holds {int} books", (Ctx ctx, Integer n) -> ctx.shelf());
    }
}

File: conformance/bundles/22-reference-ambiguous-anchor/library.steps.cs
// C# sibling of library.steps.ts / .rs (bundle 22-reference-ambiguous-anchor).
using Varar;
using Varar.Core;

namespace Varar.Corpus.B22;

public static class LibrarySteps
{
    public static void Register(Steps s)
    {
        s.Stimulus(
            "I shelve {int} books",
            (state, n) => Value.Map([new("shelf", Value.Of(ShelfOf(state) + AsLong(n)))]));

        s.Stimulus(
            "I borrow a book",
            state => Value.Map([new("shelf", Value.Of(ShelfOf(state) - 1))]));

        s.Sensor("The shelf holds {int} books", (state, n) => Value.Of(ShelfOf(state)));
    }

    public static Value State() => Value.Map([new("shelf", Value.Of(0))]);

    private static long ShelfOf(Value state) =>
        state is VMap m && m.Entries.TryGetValue("shelf", out var v) && v is VInt i ? i.Int : 0;

    private static long AsLong(Value v) => v is VInt i ? i.Int : 0;
}
